In [1]:
from pyspark.sql import SparkSession
from pyspark.sql.functions import * 
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from urllib.request import urlretrieve
from owslib.wfs import WebFeatureService
from dotenv import load_dotenv
from pyspark.sql import functions as F
from pyspark.sql import Window

# Create a spark session
spark = (
    SparkSession.builder.appName("analysis")
    .config("spark.sql.repl.eagerEval.enabled", True) 
    .config("spark.sql.parquet.cacheMetadata", "true")
    .config("spark.sql.session.timeZone", "Etc/UTC")
    .config("spark.driver.memory", "2g")
    .config("spark.executor.memory", "6g")
    .getOrCreate()
)

22/10/03 21:18:10 WARN Utils: Your hostname, DESKTOP-JJJD94T resolves to a loopback address: 127.0.1.1; using 192.168.180.114 instead (on interface eth0)
22/10/03 21:18:10 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address


Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


22/10/03 21:18:11 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable


In [2]:
sdf = spark.read.parquet("../data/curated/finaldf.parquet/")
sdf.limit(5)

+--------------+------------+-------+--------+-----------+------------------+--------------------+-----+------+--------------------+--------------------+---------+------------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+----------------+----------------+
|order_datetime|merchant_abn|user_id|postcode|consumer_id|      dollar_value|            order_id|state|gender|        company_name|                tags|take_rate|revenue_band|total_pop|under10_pop|adolsc_pop|yng_adult_pop|mid_age_pop|old_pop|num_earners|median_age|median_income|mean_income|consumer_fraud_%|merchant_fraud_%|
+--------------+------------+-------+--------+-----------+------------------+--------------------+-----+------+--------------------+--------------------+---------+------------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+----------------+----------------+
|    2022-07-13| 49891706470|  22131|    6731|      58911|25.782245737474312|90498b80-984f-43c...|   WA|  Male|Non Vestibulum In...|tent and awning s...|     5.80|           a|     8374|       1373|      1185|         2200|       2600|   1016|       2907|        39|        48034|      56306|             0.1|             0.1|
|    2021-08-19| 96680767841|  22131|    6731|      58911|    415.9547984625|829843bc-c571-493...|   WA|  Male|      Ornare Limited|motor vehicle sup...|     5.91|           a|     8374|       1373|      1185|         2200|       2600|   1016|       2907|        39|        48034|      56306|             0.1|             0.1|
|    2022-07-13| 43186523025|  22131|    6731|      58911|33.082322196774484|c12d2520-12c8-40f...|   WA|  Male|Lorem Ipsum Sodal...|florists supplies...|     4.47|           b|     8374|       1373|      1185|         2200|       2600|   1016|       2907|        39|        48034|      56306|             0.1|             0.1|
|    2021-08-19| 66370248931|  22131|    6731|      58911| 82.54046571771035|cc8ac5c5-0dfc-4e1...|   WA|  Male|        Morbi Non PC|cable, satellite,...|     3.15|           b|     8374|       1373|      1185|         2200|       2600|   1016|       2907|        39|        48034|      56306|             0.1|             0.1|
|    2022-05-06| 89726005175|  22131|    6731|      58911| 58.46693553052902|b1f179f8-2c15-4f5...|   WA|  Male| Est Nunc Consulting|tent and awning s...|     6.01|           a|     8374|       1373|      1185|         2200|       2600|   1016|       2907|        39|        48034|      56306|             0.1|             0.1|
+--------------+------------+-------+--------+-----------+------------------+--------------------+-----+------+--------------------+--------------------+---------+------------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+----------------+----------------+

In [3]:
sdf = sdf.withColumns({
    "year": date_format(col("order_datetime"), "yyyy").cast("long"),
    "month": date_format(col("order_datetime"), "MM").cast("long"),
    "day": date_format(col("order_datetime"), "dd").cast("long")
})

sdf.limit(5)

22/10/03 21:18:32 WARN package: Truncated the string representation of a plan since it was too large. This behavior can be adjusted by setting 'spark.sql.debug.maxToStringFields'.


+--------------+------------+-------+--------+-----------+------------------+--------------------+-----+------+--------------------+--------------------+---------+------------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+----------------+----------------+----+-----+---+
|order_datetime|merchant_abn|user_id|postcode|consumer_id|      dollar_value|            order_id|state|gender|        company_name|                tags|take_rate|revenue_band|total_pop|under10_pop|adolsc_pop|yng_adult_pop|mid_age_pop|old_pop|num_earners|median_age|median_income|mean_income|consumer_fraud_%|merchant_fraud_%|year|month|day|
+--------------+------------+-------+--------+-----------+------------------+--------------------+-----+------+--------------------+--------------------+---------+------------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+----------------+----------------+----+-----+---+
|    2022-07-13| 49891706470|  22131|    6731|      58911|25.782245737474312|90498b80-984f-43c...|   WA|  Male|Non Vestibulum In...|tent and awning s...|     5.80|           a|     8374|       1373|      1185|         2200|       2600|   1016|       2907|        39|        48034|      56306|             0.1|             0.1|2022|    7| 13|
|    2021-08-19| 96680767841|  22131|    6731|      58911|    415.9547984625|829843bc-c571-493...|   WA|  Male|      Ornare Limited|motor vehicle sup...|     5.91|           a|     8374|       1373|      1185|         2200|       2600|   1016|       2907|        39|        48034|      56306|             0.1|             0.1|2021|    8| 19|
|    2022-07-13| 43186523025|  22131|    6731|      58911|33.082322196774484|c12d2520-12c8-40f...|   WA|  Male|Lorem Ipsum Sodal...|florists supplies...|     4.47|           b|     8374|       1373|      1185|         2200|       2600|   1016|       2907|        39|        48034|      56306|             0.1|             0.1|2022|    7| 13|
|    2021-08-19| 66370248931|  22131|    6731|      58911| 82.54046571771035|cc8ac5c5-0dfc-4e1...|   WA|  Male|        Morbi Non PC|cable, satellite,...|     3.15|           b|     8374|       1373|      1185|         2200|       2600|   1016|       2907|        39|        48034|      56306|             0.1|             0.1|2021|    8| 19|
|    2022-05-06| 89726005175|  22131|    6731|      58911| 58.46693553052902|b1f179f8-2c15-4f5...|   WA|  Male| Est Nunc Consulting|tent and awning s...|     6.01|           a|     8374|       1373|      1185|         2200|       2600|   1016|       2907|        39|        48034|      56306|             0.1|             0.1|2022|    5|  6|
+--------------+------------+-------+--------+-----------+------------------+--------------------+-----+------+--------------------+--------------------+---------+------------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+----------------+----------------+----+-----+---+

In [4]:
sdf.select('user_id').distinct().count()

20906

20,906 total users

Get number of consumers and total number of transactions for each merchant

In [5]:
sdf.groupBy("merchant_abn").agg(
    F.count_distinct("consumer_id").alias("consumer_count"),
    F.count("consumer_id").alias("transactions_count")
).limit(5)

+------------+--------------+------------------+
|merchant_abn|consumer_count|transactions_count|
+------------+--------------+------------------+
| 73256306726|          4045|              4522|
| 83412691377|          9338|             12361|
| 38700038932|          5080|              5838|
| 38986645707|            38|                38|
| 12516851436|           183|               183|
+------------+--------------+------------------+

For each merchant, get the postcodes and the number of transactions originating from that postcode

In [6]:
merch_postcode = sdf.groupby('merchant_abn', 'postcode').count()
merch_postcode.orderBy("merchant_abn").limit(5)

+------------+--------+-----+
|merchant_abn|postcode|count|
+------------+--------+-----+
| 10023283211|    4559|    1|
| 10023283211|    5168|    1|
| 10023283211|    3415|    2|
| 10023283211|    2019|    1|
| 10023283211|    6461|    1|
+------------+--------+-----+

For each merchant, get the most frequent postcodes

In [7]:
w = Window.partitionBy('merchant_abn')

merch_postcode = merch_postcode \
    .withColumn('maxCount', F.max('count').over(w)) \
    .where(F.col('count') == F.col('maxCount')) \
    .drop('maxCount', 'count')

merch_postcode.limit(5)

+------------+--------+
|merchant_abn|postcode|
+------------+--------+
| 10023283211|    3275|
| 10023283211|    2388|
| 10023283211|    5582|
| 10342410215|    6237|
| 10342410215|    6335|
+------------+--------+

In [8]:
external_sdf = spark.read.parquet("../data/curated/externaldata.parquet/")
external_sdf.limit(5)

+--------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+
|postcode|total_pop|under10_pop|adolsc_pop|yng_adult_pop|mid_age_pop|old_pop|num_earners|median_age|median_income|mean_income|
+--------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+
|    0200|     2875|          6|      1528|         1292|         47|      2|        548|        23|         9306|      16835|
|    0800|     7679|        474|       325|         3322|       2652|    906|       5909|        33|        60937|      87791|
|    0801|     7679|        474|       325|         3322|       2652|    906|       5909|        33|        60937|      87791|
|    0804|     2980|        350|       282|          801|       1047|    500|       1873|        40|        75219|      98872|
|    0810|     2657|        326|       274|          541|        845|    386|       1479|        39|        58753|      67299|
+--------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+

In [9]:
print(merch_postcode.count())
merch_info = merch_postcode.join(external_sdf, "postcode")
print(merch_info.count())

33745


33745


In [10]:
merch_info.limit(5)

+--------+------------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+
|postcode|merchant_abn|total_pop|under10_pop|adolsc_pop|yng_adult_pop|mid_age_pop|old_pop|num_earners|median_age|median_income|mean_income|
+--------+------------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+
|    3275| 10023283211|     9963|       1127|      1317|         1319|       3117|   3083|       5765|        48|        43136|      52080|
|    2388| 10023283211|     4656|        629|       601|          751|       1463|   1212|       2656|        46|        43821|      51530|
|    5582| 10023283211|     4209|        330|       348|          454|       1176|   1901|       2021|        53|        36420|      46556|
|    6237| 10342410215|     3131|        315|       543|          341|       1139|    793|       1915|        47|        47678|      58959|
|    6335| 10342410215|     2958|        479|       295|          525|       1014|    645|       1801|        44|        58087|      78139|
+--------+------------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+

In [11]:
merch_info = merch_info.groupBy("merchant_abn").agg(
    F.round(F.avg("total_pop"), 2).alias("total_pop"),
    F.round(F.avg("under10_pop"), 2).alias("under10_pop"),
    F.round(F.avg("adolsc_pop"), 2).alias("adolsc_pop"),
    F.round(F.avg("yng_adult_pop"), 2).alias("yng_adult_pop"),
    F.round(F.avg("mid_age_pop"), 2).alias("mid_age_pop"),
    F.round(F.avg("old_pop"), 2).alias("old_pop"),
    F.round(F.avg("num_earners"), 2).alias("num_earners"),
    F.round(F.avg("median_age"), 2).alias("median_age"),
    F.round(F.avg("median_income"), 2).alias("median_income"),
    F.round(F.avg("mean_income"), 2).alias("mean_income")
)

merch_info.limit(5)

+------------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+
|merchant_abn|total_pop|under10_pop|adolsc_pop|yng_adult_pop|mid_age_pop|old_pop|num_earners|median_age|median_income|mean_income|
+------------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+
| 10023283211|   6276.0|     695.33|    755.33|       841.33|    1918.67|2065.33|    3480.67|      49.0|     41125.67|   50055.33|
| 10342410215|  4665.67|      534.0|    589.33|        604.0|     1668.0| 1208.0|    2670.33|      46.0|     50693.33|   65035.67|
| 10346855916| 11618.86|    1405.29|   1509.57|       1881.0|    3777.14|3045.86|    5949.71|      46.0|     48571.43|    61045.0|
| 10385163239|  26723.0|     2652.0|    2696.0|       3439.0|     7889.0|10047.0|     4514.0|      43.0|      38060.0|    47392.0|
| 10648956813|   7784.0|      772.0|     969.5|        999.0|     2407.0| 2636.5|     4093.5|      47.5|      39231.0|    48572.5|
+------------+---------+-----------+----------+-------------+-----------+-------+-----------+----------+-------------+-----------+